lib/simple_unique_jobs/lock.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require "murmurhash3"
+module SimpleUniqueJobs
+  class Lock
+    KEY_PREFIX = "unique"
+    KEY_PATTERN = "%<prefix>s:%<type>s:%<classname>s:%<hash>s"
+
+    def initialize(job, redis_pool)
+      @job = job
+      @redis_pool = redis_pool
+    end
+
+    def if_enqueueable
+      return if queued_unique_for > 0 && !lock("q", queued_unique_for)
+
+      yield
+    end
+
+    def unlock_enqueue
+      unlock("q")
+    end
+
+    def if_runnable
+      return yield if running_unique_for == 0
+      return unless lock("r", running_unique_for)
```

Comment:
guys, seems you are adding the Redlock into this gem to achieve the uniqueness of the sidekiq jobs. I have worked on similar tasks. I think maybe you can consider other options:

1. use https://github.com/mhenrixon/sidekiq-unique-jobs. This gem works well, the company I'm working for uses this gem. Our site has quite some traffic and this gem is working fine.
2. implement the distributed locking mechanism out side of this sidekiq job specific gem. There are a lot of places outside of the jobs that also need to avoid race conditions. So, if you implement a Redlock(whether into a separate gem or just a service in code base), you will be able to use the lock for sidekiq jobs and for anything else. 

BTW, I'm a developer based in San Francisco and I have been working with Rails for 10+ years. I'm starting a side business selling BDSM product. Yesterday, when I was trying to run ads on Fetlife, I found you guys have some open source repos and you are a Ruby shop! Haha! so, I can't help participate this code review. 😄 The site you built is great. Please keep it up. 

Greetings from a Ruby developer + BDSM fan from San Francisco!

Reply:
@TheChineseDude thanks for this!

Appreciate this gem looks like reinventing wheels. FWIW:

- `sidekiq-unique-jobs` is fantastic, however parts of it don't scale well (they have _O(n)_ or even _O(n log n)_ cost again the size if the scheduled queue, for instance reaping; we need _O(1)_)
- Redlock is great (and we use it!) but we don't need the strict guarantees it provides in this use case. Also we don't _exactly_ need a lock, more of a concurrent flag the a given job can be dropped/skipped (ie. there isn't a critical zone to be run within a lock while other threads wait).

Hence the much more lightweight implementation.